File: web/gee_runner/src/runner.js
import fs from 'node:fs/promises';
import path from 'node:path';
import vm from 'node:vm';
import ee from '@google/earthengine';
import { google } from 'googleapis';

const TASK_COMPLETE = new Set(['COMPLETED', 'FAILED', 'CANCELLED']);

/** 检查资产的实际类型 */
function checkAssetType(assetId) {
  return new Promise((resolve, reject) => {
    ee.data.getAsset(assetId, (result) => {
      if (result && result.type) {
        resolve(result.type);
      } else {
        reject(new Error(`Asset '${assetId}' was not found or the service account has no access.`));
      }
    }, (error) => {
      reject(new Error(`Asset '${assetId}' was not found or the service account has no access: ${error?.message || error}`));
    });
  });
}

/** ImageCollection 专有方法（在 Image 上不可用） */
const IMAGE_COLLECTION_METHODS = [
  'filterDate', 'filterBounds', 'filter', 'sort', 'limit',
  'mosaic', 'composite', 'reduceColumns', 'getRegion',
  'mean', 'median', 'min', 'max', 'sum', 'count',
  'select', 'aggregate_array', 'aggregate_total'
];

/** 预处理源代码，验证资产类型并自动修复不匹配的类型 */
async function preprocessSource(source) {
  const fixes = [];
  const assetChecks = [];

  // 查找所有 ee.ImageCollection('asset_id') 调用（包括链式调用）
  // 同时捕获赋值给的变量名
  const imageCollectionAssignRegex = /(?:var|let|const)?\s*(\w+)\s*=\s*ee\.ImageCollection\s*\(\s*['"]([^'"]+)['"]\s*\)/g;
  let match;
  const imageCollectionVars = {};  // 变量名 -> 资产ID
  
  while ((match = imageCollectionAssignRegex.exec(source)) !== null) {
    const varName = match[1];
    const assetId = match[2];
    imageCollectionVars[varName] = assetId;
    assetChecks.push({ assetId, currentType: 'ImageCollection', fullMatch: match[0] });
  }

  // 也查找直接调用（不带变量赋值）
  const imageCollectionRegex = /ee\.ImageCollection\s*\(\s*['"]([^'"]+)['"]\s*\)/g;
  while ((match = imageCollectionRegex.exec(source)) !== null) {
    const assetId = match[1];
    if (!assetChecks.find(c => c.assetId === assetId)) {
      assetChecks.push({ assetId, currentType: 'ImageCollection', fullMatch: match[0] });
    }
  }

  // 查找所有 ee.Image('asset_id') 调用
  const imageRegex = /ee\.Image\s*\(\s*['"]([^'"]+)['"]\s*\)/g;
  while ((match = imageRegex.exec(source)) !== null) {
    const assetId = match[1];
    if (!assetChecks.find(c => c.assetId === assetId)) {
      assetChecks.push({ assetId, currentType: 'Image', fullMatch: match[0] });
    }
  }

  if (assetChecks.length === 0) {
    return { source, fixes };
  }

  // 批量检查资产类型
  console.log(`[ASSET] 检测到 ${assetChecks.length} 个资产引用，正在验证类型...`);
  const checkedTypes = await Promise.all(
    assetChecks.map(async ({ assetId }) => ({
      assetId,
      actualType: await checkAssetType(assetId)
    }))
  );

  // 构建资产类型映射
  const typeMap = {};
  checkedTypes.forEach(({ assetId, actualType }) => {
    if (actualType) {
      typeMap[assetId] = actualType;
    }
  });

  // 应用修复
  let fixedSource = source;
  const varsToFix = {};  // 需要修复的变量名 -> 资产ID

  assetChecks.forEach(({ assetId, currentType, fullMatch }) => {
    const actualType = typeMap[assetId];
    if (!actualType) return;

    // 如果当前是 ImageCollection 但实际是 Image
    if (currentType === 'ImageCollection' && actualType === 'Image') {
      // 找到完整的链式调用
      const chainRegex = new RegExp(
        `ee\\.ImageCollection\\s*\\(\\s*['"]${assetId.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}['"]\\s*\\)(\\s*\\.\\s*[\\w.]+\\s*\\([^)]*\\))*`,
        'g'
      );
      
      fixedSource = fixedSource.replace(chainRegex, (fullChain) => {
        let newCall = `ee.Image('${assetId}')`;
        
        // 找到所有链式方法调用
        const escapedId = assetId.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
        const baseCallRegex = new RegExp(`ee\\.ImageCollection\\s*\\(\\s*['"]${escapedId}['"]\\s*\\)`);
        const methodChain = fullChain
          .replace(baseCallRegex, '')
          .match(/\.(\w+)\s*\([^)]*\)/g) || [];
        
        let clipArgs = null;
        
        // 处理每个方法
        methodChain.forEach((methodCall) => {
          const methodName = methodCall.match(/\.(\w+)/)[1];
          
          if (methodName === 'filterBounds') {
            const argsMatch = methodCall.match(/filterBounds\s*\(([^)]*)\)/);
            if (argsMatch) {
              clipArgs = argsMatch[1];
            }
          }
          // 移除所有 ImageCollection 专有方法和 mosaic/composite
          // 保留可能在 Image 上也可用的方法（如 clip, select 等）
          if (!IMAGE_COLLECTION_METHODS.includes(methodName) && methodName !== 'mosaic' && methodName !== 'composite') {
            newCall += methodCall;
          }
        });
        
        if (clipArgs) {
          newCall += `.clip(${clipArgs})`;
        }
        
        return newCall;
      });

      // 记录需要修复后续使用的变量
      Object.entries(imageCollectionVars).forEach(([varName, varAssetId]) => {
        if (varAssetId === assetId) {
          varsToFix[varName] = assetId;
        }
      });

      fixes.push({
        assetId,
        original: fullMatch,
        fixed: `ee.Image('${assetId}')`,
        reason: `资产 '${assetId}' 实际是 Image 类型，不是 ImageCollection。已自动转换为 Image 并移除不兼容的方法链。`
      });
      console.log(`[ASSET] 修复: ${assetId} (ImageCollection -> Image)`);
    }
    // 如果当前是 Image 但实际是 ImageCollection
    else if (currentType === 'Image' && actualType === 'ImageCollection') {
      const newCall = `ee.ImageCollection('${assetId}')`;
      fixedSource = fixedSource.replace(fullMatch, newCall);
      fixes.push({
        assetId,
        original: fullMatch,
        fixed: newCall,
        reason: `资产 '${assetId}' 实际是 ImageCollection 类型，不是 Image`
      });
      console.log(`[ASSET] 修复: ${assetId} (Image -> ImageCollection)`);
    }
  });

  // 修复变量后续使用中的 ImageCollection 专有方法
  Object.entries(varsToFix).forEach(([varName, assetId]) => {
    // 找到所有对该变量的 ImageCollection 方法调用并移除
    // 例如: varName.mosaic() -> varName
    IMAGE_COLLECTION_METHODS.forEach(method => {
      const methodRegex = new RegExp(`\\b${varName.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\s*\\.\\s*${method}\\s*\\([^)]*\\)\\s*\\.\\s*`, 'g');
      fixedSource = fixedSource.replace(methodRegex, `${varName}.`);
      
      // 也处理链式末端的情况
      const endRegex = new RegExp(`\\b${varName.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\s*\\.\\s*${method}\\s*\\([^)]*\\)\\s*;`, 'g');
      fixedSource = fixedSource.replace(endRegex, `${varName};`);
      
      // 处理变量直接赋值给另一个变量的情况
      const assignRegex = new RegExp(`\\b${varName.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\s*\\.\\s*${method}\\s*\\([^)]*\\)`, 'g');
      fixedSource = fixedSource.replace(assignRegex, varName);
    });
    
    // 处理 mosaic/composite
    ['mosaic', 'composite'].forEach(method => {
      const regex = new RegExp(`\\b${varName.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\s*\\.\\s*${method}\\s*\\([^)]*\\)`, 'g');
      fixedSource = fixedSource.replace(regex, varName);
    });
  });

  return { source: fixedSource, fixes };
}

function authenticate({ auth, keyFile }) {
  return new Promise((resolve, reject) => {
    const timeout = setTimeout(() => {
      reject(new Error('Authentication timed out after 60 seconds. Check HTTPS_PROXY and network access to Google.'));
    }, 60000);
    const finish = (callback) => (...args) => {
      clearTimeout(timeout);
      callback(...args);
    };
    const initialized = () => ee.initialize(null, null, finish(resolve), finish(reject));
    if (auth === 'service-account') {
      if (!keyFile) return finish(reject)(new Error('Service-account mode requires --key or GEE_PRIVATE_KEY_FILE.'));
      fs.readFile(path.resolve(keyFile), 'utf8')
        .then(async (contents) => {
          const privateKey = JSON.parse(contents);
          const scopes = [
            'https://www.googleapis.com/auth/earthengine',
            'https://www.googleapis.com/auth/cloud-platform',
            'https://www.googleapis.com/auth/devstorage.full_control',
          ];
          const client = new google.auth.JWT(
            privateKey.client_email, null, privateKey.private_key, scopes
          );
          const refresh = async () => {
            const token = await client.authorize();
            console.log('Service-account token acquired.');
            return {
              access_token: token.access_token,
              token_type: token.token_type || 'Bearer',
              expires_in: (token.expiry_date - Date.now()) / 1000,
            };
          };
          const token = await refresh();
          console.log('Initializing Earth Engine API...');
          ee.data.setAuthTokenRefresher((_, callback) => {
            refresh().then(callback).catch((error) => callback({ error }));
          });
          ee.data.setAuthToken(
            privateKey.client_email, token.token_type, token.access_token, token.expires_in,
            scopes, initialized, false
          );
        })
        .catch((error) => {
          const code = error?.code || error?.cause?.code || 'unknown';
          const proxy = process.env.HTTPS_PROXY || process.env.HTTP_PROXY || '未配置';
          finish(reject)(new Error(`Google OAuth 认证失败（${code}）：${error?.message || error}。Token 地址：https://oauth2.googleapis.com/token；当前代理：${proxy}。请检查网络或在 web/gee_runner/.env 配置 HTTPS_PROXY。`));
        });
      return;
    }
    finish(reject)(new Error(
      'OAuth is not supported by the Earth Engine Node.js SDK. Use a service account: set GEE_AUTH=service-account and GEE_PRIVATE_KEY_FILE=./service-account.json.'
    ));
  });
}

function makeExport(tasks) {
  const submit = (factory) => (config) => {
    const task = factory(config);
    task.start();
    tasks.push(task);
    console.log(`Export started: ${config.description || task.id}`);
    return task;
  };
  return {
    image: { toDrive: submit(ee.batch.Export.image.toDrive), toAsset: submit(ee.batch.Export.image.toAsset), toCloudStorage: submit(ee.batch.Export.image.toCloudStorage) },
    table: { toDrive: submit(ee.batch.Export.table.toDrive), toAsset: submit(ee.batch.Export.table.toAsset), toCloudStorage: submit(ee.batch.Export.table.toCloudStorage) },
  };
}

export async function runScript(scriptPath, options) {
  console.log('Authenticating with Google Earth Engine...');
  await authenticate(options);
  console.log('Authentication complete. Running script...');
  const absolutePath = path.resolve(scriptPath);
  const rawSource = await fs.readFile(absolutePath, 'utf8');
  
  // 预处理代码，验证并修复资产类型
  const { source: processedSource, fixes } = await preprocessSource(rawSource);
  if (fixes.length > 0) {
    console.log(`[ASSET] 自动修复了 ${fixes.length} 个资产类型问题:`);
    fixes.forEach((fix, i) => {
      console.log(`  ${i + 1}. ${fix.reason}`);
    });
    console.log('[ASSET] 已应用资产类型修复；最终代码仅显示在成功结果面板。');
  }
  
  const tasks = [];
  const context = vm.createContext({
    ee,
    Export: makeExport(tasks),
    console,
    print: (...values) => console.log(...values),
    Map: { addLayer: (...values) => console.log('Map.addLayer:', ...values), centerObject: () => {}, setCenter: () => {} },
  });
  new vm.Script(processedSource, { filename: absolutePath }).runInContext(context);
  return { tasks, processedSource };
}

export async function runSource(source, filename, options) {
  console.log('Authenticating with Google Earth Engine...');
  await authenticate(options);
  console.log('Authentication complete. Running JavaScript...');
  
  // 预处理代码，验证并修复资产类型
  const { source: processedSource, fixes } = await preprocessSource(source);
  if (fixes.length > 0) {
    console.log(`[ASSET] 自动修复了 ${fixes.length} 个资产类型问题:`);
    fixes.forEach((fix, i) => {
      console.log(`  ${i + 1}. ${fix.reason}`);
    });
    console.log('[ASSET] 已应用资产类型修复；最终代码仅显示在成功结果面板。');
  }
  
  const tasks = [];
  const layerChecks = [];
  const verifyLayer = (image, name) => new Promise((resolve, reject) => {
    image.getMap({ min: 0, max: 1 }, (map, error) => {
      if (error) reject(new Error(`Layer '${name || 'unnamed'}' failed validation: ${error}`));
      else resolve(map);
    });
  });
  const context = vm.createContext({
    ee, Export: makeExport(tasks), console,
    print: (...values) => console.log('GEE_RESULT:', ...values),
    Map: {
      addLayer: (image, vis, name) => {
        console.log('GEE_RESULT: Map.addLayer:', name || 'unnamed');
        layerChecks.push(verifyLayer(image, name));
      },
      centerObject: () => {}, setCenter: () => {},
    },
  });
  new vm.Script(processedSource, { filename }).runInContext(context);
  await Promise.all(layerChecks);
  return { tasks, processedSource };
}

export async function waitForTasks(tasks, intervalMs = 10000) {
  const pending = new Set(tasks.map((task) => task.id));
  while (pending.size) {
    const states = await Promise.all([...pending].map((id) => new Promise((resolve, reject) => ee.data.getTaskStatus([id], (result) => resolve(result[0]), reject))));
    for (const state of states) {
      console.log(`${state.id}: ${state.state}${state.error_message ? ` (${state.error_message})` : ''}`);
      if (TASK_COMPLETE.has(state.state)) pending.delete(state.id);
    }
    if (pending.size) await new Promise((resolve) => setTimeout(resolve, intervalMs));
  }
}

File: web/gee_runner/runner.py
"""Run Google Earth Engine Python scripts with service-account credentials."""

from __future__ import annotations

import argparse
import json
import os
import runpy
import sys
import time
import traceback
import urllib.request
import urllib.error
import socket
import http.client
import random
from pathlib import Path


def load_env(env_file: Path) -> None:
    if not env_file.exists():
        return
    for line in env_file.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        os.environ.setdefault(key.strip(), value.strip())


def load_codex_config() -> None:
    """Read the user's Codex-compatible provider config without importing TOML dependencies."""
    config_file = Path(os.environ.get("CODEX_CONFIG_FILE", Path.home() / ".codex" / "config.toml"))
    if config_file.exists():
        section = None
        for raw in config_file.read_text(encoding="utf-8").splitlines():
            line = raw.strip()
            if not line or line.startswith("#"):
                continue
            if line.startswith("[") and line.endswith("]"):
                section = line[1:-1].strip()
                continue
            if "=" not in line:
                continue
            key, value = (part.strip() for part in line.split("=", 1))
            if value.startswith('"') and value.endswith('"'):
                value = value[1:-1]
            if section == "model_providers.llmapi":
                if key == "base_url":
                    os.environ.setdefault("OPENAI_BASE_URL", value)
                elif key == "model":
                    os.environ.setdefault("OPENAI_MODEL", value)
    auth_file = Path(os.environ.get("CODEX_AUTH_FILE", Path.home() / ".codex" / "auth.json"))
    if auth_file.exists() and not os.environ.get("OPENAI_API_KEY"):
        data = json.loads(auth_file.read_text(encoding="utf-8"))
        key = data.get("OPENAI_API_KEY") or data.get("api_key") or data.get("key")
        if key:
            os.environ["OPENAI_API_KEY"] = key


def initialize(key_file: str):
    import ee

    key_path = Path(key_file).resolve()
    if not key_path.is_file():
        raise FileNotFoundError(f"Service-account key not found: {key_path}")
    key = json.loads(key_path.read_text(encoding="utf-8"))
    service_account = key.get("client_email")
    if not service_account:
        raise ValueError("The key file does not contain client_email.")
    project = os.environ.get("GEE_PROJECT") or key.get("project_id")
    print("Authenticating with Google Earth Engine...")
    ee.Initialize(ee.ServiceAccountCredentials(service_account, str(key_path)), project=project)
    print("Authentication complete. Running script...")
    return ee


def wait_for_tasks(tasks, interval: int) -> None:
    pending = set(tasks)
    terminal = {"COMPLETED", "FAILED", "CANCELLED", "CANCEL_REQUESTED"}
    while pending:
        for task in list(pending):
            status = task.status()
            state = status.get("state", "UNKNOWN")
            detail = status.get("error_message", "")
            print(f"{status.get('id', 'unknown')}: {state}{f' ({detail})' if detail else ''}")
            if state == "FAILED":
                raise RuntimeError(detail or f"Earth Engine task {status.get('id')} failed")
            if state in terminal:
                pending.remove(task)
        if pending:
            time.sleep(interval)


def call_fix_api(source: str, error: str, output_format: str = "py") -> str:
    """调用 AI API 修复/转换 GEE 代码。output_format: 'js' 返回 JavaScript, 'py' 返回 Python。"""
    api_key = os.environ.get("OPENAI_API_KEY")
    if not api_key:
        raise RuntimeError("Run failed and OPENAI_API_KEY is not configured; cannot auto-fix.")
    model = os.environ.get("OPENAI_MODEL", "gpt-5.4")
    if output_format == "js":
        prompt = (
            "You are preparing a Google Earth Engine JavaScript script for server execution. Return ONLY complete corrected GEE JavaScript code (the same API style as the Earth Engine Code Editor), with no Markdown fences or explanation. Fix bugs and preserve the user's intent and export configuration.\n\n"
            f"Error:\n{error}\n\nCurrent script:\n{source}"
        )
    else:
        prompt = (
            "You are preparing a Google Earth Engine script for server execution. Return ONLY complete corrected Python Earth Engine API source code, with no Markdown fences or explanation. The input may be JavaScript from the Earth Engine Code Editor; if it is JavaScript, convert it to equivalent Python API code. Preserve the user's intent and export configuration.\n\n"
            f"Error:\n{error}\n\nCurrent script:\n{source}"
        )
    base_url = os.environ.get("OPENAI_BASE_URL", "https://api.openai.com").rstrip("/")
    if not base_url.endswith("/v1"):
        base_url += "/v1"
    headers = {"Authorization": f"Bearer {api_key}", "Content-Type": "application/json", "User-Agent": "gee-runner/1.0", "Connection": "close"}
    attempts = [("responses", {"model": model, "input": prompt, "max_output_tokens": 12000}), ("chat/completions", {"model": model, "messages": [{"role": "user", "content": prompt}], "max_tokens": 12000})]
    failures = []
    result = None
    transient_retries = max(1, int(os.environ.get("OPENAI_RETRIES", "3")))
    for endpoint, body in attempts:
        for retry in range(transient_retries):
            request = urllib.request.Request(base_url + "/" + endpoint, data=json.dumps(body).encode("utf-8"), headers=headers, method="POST")
            try:
                with urllib.request.urlopen(request, timeout=180) as response:
                    result = json.loads(response.read().decode("utf-8"))
                break
            except urllib.error.HTTPError as error:
                detail = error.read().decode('utf-8', errors='replace')
                transient = error.code in (408, 409, 425, 429, 500, 502, 503, 504) or "stream disconnected" in detail.lower() or "stream closed" in detail.lower()
                failures.append(f"{endpoint}: HTTP {error.code}: {detail}")
                if transient and retry + 1 < transient_retries:
                    delay = min(20, 2 ** retry + random.random())
                    print(f"AI API temporary failure ({error.code}), retrying {retry + 1}/{transient_retries - 1} in {delay:.1f}s...", flush=True)
                    time.sleep(delay)
                    continue
            except (urllib.error.URLError, ConnectionError, TimeoutError, socket.timeout, http.client.HTTPException) as error:
                failures.append(f"{endpoint}: connection error: {error}")
                if retry + 1 < transient_retries:
                    delay = min(20, 2 ** retry + random.random())
                    print(f"AI API connection failure, retrying {retry + 1}/{transient_retries - 1} in {delay:.1f}s...", flush=True)
                    time.sleep(delay)
                    continue
            break
        if result is not None:
            break
    if result is None:
        raise RuntimeError("OpenAI API failed. " + " | ".join(failures))
    text = result.get("output_text") or ""
    if not text and result.get("choices"):
        text = result["choices"][0].get("message", {}).get("content", "")
    if not text:
        for item in result.get("output", []):
            for content in item.get("content", []):
                if content.get("type") == "output_text":
                    text += content.get("text", "")
    if not text:
        raise RuntimeError("OpenAI API returned no corrected source code.")
    if text.startswith("```"):
        text = text.split("\n", 1)[1] if "\n" in text else text
        text = text.rsplit("```", 1)[0]
    return text.strip() + "\n"


def run_once(source: str, filename: str, key_file: str, wait: bool, poll_seconds: int) -> None:
    ee = initialize(key_file)
    submitted = []
    original_start = ee.batch.Task.start

    def start_and_track(task):
        original_start(task)
        submitted.append(task)
        print(f"Export started: {task.status().get('id', 'unknown')}")

    ee.batch.Task.start = start_and_track
    namespace = {"ee": ee, "Export": ee.batch.Export}
    exec(compile(source, filename, "exec"), namespace, namespace)
    if submitted and wait:
        wait_for_tasks(submitted, poll_seconds)


def is_javascript(source: str) -> bool:
    markers = ("//", "var ", "let ", "const ", "function ", "Map.")
    return any(line.lstrip().startswith(markers) for line in source.splitlines() if line.strip())


def _run_js_once(source: str, key_file: str, wait: bool) -> tuple[str, bool, str]:
    """通过子进程执行 JS runner，返回 (最终代码, 是否成功, 输出日志)。"""
    import subprocess as _sp
    root = Path(__file__).resolve().parent
    command = ["node", str(root / "bin" / "gee-js-runner.js")]
    if wait:
        command.append("--wait")
    env = os.environ.copy()
    env["NODE_OPTIONS"] = env.get("NODE_OPTIONS", "")
    env["GEE_PRIVATE_KEY_FILE"] = key_file
    _fix_proxy_env(env)
    process = _sp.Popen(command, cwd=root, stdin=_sp.PIPE, stdout=_sp.PIPE, stderr=_sp.STDOUT, text=True, encoding="utf-8", errors="replace", env=env)
    process.stdin.write(source)
    process.stdin.close()
    all_lines = []
    skip_block = False
    for line in process.stdout:
        stripped = line.strip()
        if stripped == "FINAL_CODE_BEGIN":
            skip_block = True
            all_lines.append(line)
            continue
        if stripped == "FINAL_CODE_END":
            skip_block = False
            all_lines.append(line)
            continue
        all_lines.append(line)
        if not skip_block:
            sys.stdout.write(line)
            sys.stdout.flush()
    code = process.wait()
    raw_output = "".join(all_lines)
    final = raw_output.split("FINAL_CODE_BEGIN\n", 1)[-1].split("\nFINAL_CODE_END", 1)[0] if "FINAL_CODE_BEGIN" in raw_output else source
    log_output = "".join(line for line in all_lines if line.strip() not in ("FINAL_CODE_BEGIN", "FINAL_CODE_END"))
    success = code == 0 and "SUCCESS" in raw_output
    return final, success, log_output


NETWORK_ERROR_PATTERNS = [
    "socket hang up",
    "aborted",
    "ECONNREFUSED",
    "ECONNRESET",
    "ENOTFOUND",
    "network is unreachable",
    "ETIMEDOUT",
    "connection reset",
    "getaddrinfo ENOTFOUND",
    "Invalid JSON: Error:",
    "SELF_SIGNED_CERT",
    "CERT_HAS_EXPIRED",
    "unable to get local issuer certificate",
    "403 Forbidden",
    "502 Bad Gateway",
    "503 Service Unavailable",
    "504 Gateway Timeout",
    "ETIMEDOUT",
    "ESOCKETTIMEDOUT",
    "EPIPE",
    "EAI_AGAIN",
    "ENETUNREACH",
    "WinError 10013",
    "access permissions",
    "request to https://oauth2.googleapis.com failed",
    "request to https://www.googleapis.com failed",
    "google oauth 认证失败",
]


def _is_network_error(output: str) -> bool:
    """判断错误是否为网络连接问题（非代码逻辑错误）。"""
    lower = output.lower()
    return any(p.lower() in lower for p in NETWORK_ERROR_PATTERNS)


def _detect_windows_proxy() -> str | None:
    """检测 Windows 系统代理设置，返回代理 URL 或 None。"""
    if sys.platform != "win32":
        return None
    try:
        import winreg
        key_path = r"Software\Microsoft\Windows\CurrentVersion\Internet Settings"
        with winreg.OpenKey(winreg.HKEY_CURRENT_USER, key_path) as key:
            enabled, _ = winreg.QueryValueEx(key, "ProxyEnable")
            if not enabled:
                return None
            server, _ = winreg.QueryValueEx(key, "ProxyServer")
            if server:
                if not server.startswith(("http://", "https://", "socks://", "socks5://")):
                    server = f"http://{server}"
                return server
    except Exception:
        pass
    return None


def _fix_proxy_env(env: dict) -> None:
    """检查并修复代理环境变量：优先使用用户配置，其次自动检测 Windows 系统代理。"""
    proxy_keys = [k for k in ("HTTPS_PROXY", "HTTP_PROXY", "ALL_PROXY", "http_proxy", "https_proxy", "all_proxy") if k in env]
    proxy_url = None
    if proxy_keys:
        proxy_url = env.get("HTTPS_PROXY") or env.get("https_proxy") or env.get("HTTP_PROXY") or env.get("http_proxy", "")
    if not proxy_url:
        detected = _detect_windows_proxy()
        if detected:
            proxy_url = detected
            env["HTTPS_PROXY"] = proxy_url
            env["HTTP_PROXY"] = proxy_url
            print(f"[PROXY] Auto-detected Windows system proxy: {proxy_url}", flush=True)
    if not proxy_url:
        print("[PROXY] No proxy configured.", flush=True)
        return
    try:
        from urllib.parse import urlparse as _urlparse
        parsed = _urlparse(proxy_url)
        host = parsed.hostname
        port = parsed.port or (443 if parsed.scheme == "https" else 80)
        if host and port:
            import socket as _socket
            sock = _socket.create_connection((host, port), timeout=2)
            sock.close()
            print(f"[PROXY] Proxy {proxy_url} is reachable.", flush=True)
            return
    except Exception:
        pass
    print(f"[PROXY] WARNING: Proxy {proxy_url} appears unreachable, but keeping proxy settings as configured.", flush=True)
    print(f"[PROXY] If GEE requests fail, ensure your proxy is running or update {proxy_url} in .env", flush=True)


def run_js_engine(source: str, key_file: str, wait: bool, max_fixes: int) -> str:
    """通过 Node.js @google/earthengine JS SDK 运行 GEE JavaScript 代码，支持自动修复。"""
    attempt = 0
    seen_sources = {source}
    unlimited_fix_limit = 12
    while True:
        attempt += 1
        limit_label = "until success or stopped" if max_fixes < 0 else str(max_fixes + 1)
        print(f"\n=== JS engine attempt {attempt}/{limit_label} ===", flush=True)
        final, success, output_text = _run_js_once(source, key_file, wait)
        if success:
            print("SUCCESS")
            return final
        error_msg = f"JS engine failed:\n{output_text}"
        print(f"\nRUN FAILED (attempt {attempt}/{limit_label})", flush=True)
        if _is_network_error(output_text):
            print("NETWORK ERROR DETECTED: 这是网络连接问题，不是代码错误。", flush=True)
            print("建议：1) 检查网络连接  2) 尝试切换到 PY 引擎（Python SDK 通常更稳定）", flush=True)
            raise RuntimeError(
                f"网络连接错误，auto-fix 无效。请检查网络或切换到 PY 引擎。\n\n{error_msg}"
            )
        if max_fixes >= 0 and attempt > max_fixes:
            raise RuntimeError(f"FAILED after {max_fixes} auto-fix attempt(s): {error_msg}")
        if max_fixes < 0 and attempt > unlimited_fix_limit:
            raise RuntimeError(f"FAILED after {unlimited_fix_limit} auto-fix attempts; unable to repair this record:\n{error_msg}")
        print(f"AUTO-FIX: sending the failure to OpenAI (repair #{attempt})...", flush=True)
        corrected = call_fix_api(source, error_msg, output_format="js")
        if not corrected.strip() or corrected in seen_sources:
            raise RuntimeError(f"FAILED: 自动修复未产生新的代码，已停止当前记录。\n{error_msg}")
        seen_sources.add(corrected)
        source = corrected
        print("AUTO-FIX: corrected JavaScript received; restarting...", flush=True)


def run_py_engine(source: str, filename: str, key_file: str, wait: bool, poll_seconds: int, max_fixes: int) -> str:
    """通过 Python earthengine-api 运行 GEE 代码（JS 输入时先自动转换为 Python）。"""
    if is_javascript(source):
        print("Detected JavaScript source. Converting to Python Earth Engine API...", flush=True)
        source = call_fix_api(source, "JavaScript detected; converting to Python Earth Engine API equivalent.", output_format="py")
        compile(source, str(filename), "exec")
        print("JavaScript converted to Python successfully.", flush=True)
    for attempt in range(max_fixes + 1):
        try:
            run_once(source, str(filename), key_file, wait, poll_seconds)
            print("SUCCESS")
            return source
        except Exception as error:
            details = traceback.format_exc()
            print(f"\nRUN FAILED (attempt {attempt + 1}/{max_fixes + 1})", flush=True)
            print(details, end="", flush=True)
            if _is_network_error(details):
                print("NETWORK ERROR DETECTED: 这是网络连接问题，不是代码错误。", flush=True)
                raise RuntimeError(
                    f"网络连接错误，auto-fix 无效。请检查网络连接后重试。\n\n{details}"
                ) from error
            if attempt >= max_fixes:
                raise RuntimeError(f"FAILED after {max_fixes} auto-fix attempt(s): {error}") from error
            print(f"AUTO-FIX: sending the failure to OpenAI (attempt {attempt + 1}/{max_fixes})...", flush=True)
            corrected = call_fix_api(source, details, output_format="py")
            compile(corrected, str(filename), "exec")
            source = corrected
            print("AUTO-FIX: corrected script saved; restarting...", flush=True)
    return source


def main() -> int:
    parser = argparse.ArgumentParser(description="Run a Google Earth Engine script.")
    parser.add_argument("script", help="Path to a .js/.py Earth Engine script")
    parser.add_argument("--engine", choices=["js", "py"], default="js", help="Execution engine: js (default) or py")
    parser.add_argument("--key", help="Service-account JSON key path")
    parser.add_argument("--wait", action="store_true", help="Wait for submitted export tasks")
    parser.add_argument("--poll-seconds", type=int, default=10)
    parser.add_argument("--max-fixes", type=int, default=3)
    parser.add_argument("--code-stdin", action="store_true", help="Read script directly from stdin")
    args = parser.parse_args()

    load_env(Path(__file__).with_name(".env"))
    load_codex_config()
    _fix_proxy_env(os.environ)
    script_path = Path(args.script).resolve() if args.script else Path("<editor>")
    key_file = args.key or os.environ.get("GEE_PRIVATE_KEY_FILE", "service-account.json")
    source = sys.stdin.read() if args.code_stdin else script_path.read_text(encoding="utf-8")
    if "\ufffd" in source:
        raise ValueError("The submitted script contains replacement characters (U+FFFD), which indicates text was corrupted before execution. Paste the original source again using UTF-8.")

    if args.engine == "js":
        print(f"=== Using JS engine ({script_path.name}) ===", flush=True)
        result = run_js_engine(source, key_file, args.wait, args.max_fixes)
    else:
        print(f"=== Using PY engine ({script_path.name}) ===", flush=True)
        result = run_py_engine(source, script_path, key_file, args.wait, args.poll_seconds, args.max_fixes)

    print("FINAL_CODE_BEGIN")
    print(result, end="" if result.endswith("\n") else "\n")
    print("FINAL_CODE_END")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as error:
        print(f"Failed: {error}", file=sys.stderr)
        raise SystemExit(1)
